src/mcpdotnet/Configuration/McpServerBuilderExtensions.Tools.cs

```diff
@@ -0,0 +1,249 @@
+    /// </summary>
+    /// <param name="builder">The builder instance.</param>
+    public static IMcpServerBuilder WithTools(this IMcpServerBuilder builder)
+    {
+        return WithToolsFromAssembly(builder, Assembly.GetCallingAssembly());
+    }
+
+    /// <summary>
+    /// Adds tools to the server.
+    /// </summary>
+    /// <param name="builder">The builder instance.</param>
+    /// <param name="toolTypes">Types with marked methods to add as tools to the server.</param>
+    public static IMcpServerBuilder WithTools(this IMcpServerBuilder builder, params Type[] toolTypes)
+    {
+        ArgumentNullException.ThrowIfNull(toolTypes);
+        if (toolTypes.Length == 0)
+            throw new ArgumentException("At least one tool type must be provided.", nameof(toolTypes));
+
+        var tools = new List<Tool>();
+        Dictionary<string, Func<RequestContext<CallToolRequestParams>, CancellationToken, Task<CallToolResponse>>> callbacks = [];
+
+        foreach (var type in toolTypes)
+        {
+            var methods = type.GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static);
+            foreach (var method in methods)
+            {
+                var attribute = method.GetCustomAttribute<McpToolAttribute>();
+                if (attribute != null)
+                {
+                    var tool = CreateTool(method, attribute);
+                    tools.Add(tool);
+
+                    callbacks.Add(tool.Name, async (request, cancellationToken) => await CallTool(request, method, cancellationToken));
+
+                    // register type because method is not static and so we need an instance
+                    if (!method.IsStatic)
+                        builder.Services.AddScoped(type);
+                }
+            }
+        }
+
+        builder.WithListToolsHandler((_, _) => Task.FromResult(new ListToolsResult() { Tools = tools }));
+
+        builder.WithCallToolHandler(async (request, cancellationToken) =>
+        {
+            if (request.Params != null && callbacks.TryGetValue(request.Params.Name, out var callback))
+                return await callback(request, cancellationToken);
+
+            throw new McpServerException($"Unknown tool: {request.Params?.Name}");
+        });
+
+        return builder;
+    }
+
+    /// <summary>
+    /// Adds types marked with the <see cref="McpToolTypeAttribute"/> attribute from the given assembly as tools to the server.
+    /// </summary>
+    /// <param name="builder">The builder instance.</param>
+    /// <param name="assembly">The assembly to load the types from. Null to get the current assembly</param>
+    public static IMcpServerBuilder WithToolsFromAssembly(this IMcpServerBuilder builder, Assembly? assembly = null)
+    {
+        assembly ??= Assembly.GetCallingAssembly();
+
+        List<Type> toolTypes = [];
+
+        foreach (var type in assembly.GetTypes())
```

Comment:
GetTypes produces all types, not just public ones. That means someone could walk up to an arbitrary assembly and this will register anything marked as a tool, even if it's internal. Is that the intent?

Reply:
Sorry, was not intended. Fixed with #54